packages/engine/paima-funnel/src/funnels/avail/utils.ts

```diff
@@ -0,0 +1,171 @@
+import type { ApiPromise } from 'avail-js-sdk';
+import type { Header as PolkadotHeader } from '@polkadot/types/interfaces/types';
+import type { SubmittedData } from '@paima/sm';
+import { base64Decode } from '@polkadot/util-crypto';
+import { BaseFunnelSharedApi } from '../BaseFunnel.js';
+import { createApi } from './createApi.js';
+
+export const GET_DATA_TIMEOUT = 10000;
+
+export type Header = PolkadotHeader;
+
+export function getSlotFromHeader(header: Header, api: ApiPromise): number {
+  const preRuntime = header.digest.logs.find(log => log.isPreRuntime)!.asPreRuntime;
+
+  const rawBabeDigest = api.createType('RawBabePreDigest', preRuntime[1]);
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const babeDigest = rawBabeDigest.toPrimitive() as unknown as any;
+
+  // the object is an enumeration, but all the variants have a slotNumber field
+  const slot = babeDigest[Object.getOwnPropertyNames(babeDigest)[0]].slotNumber;
+  return slot;
+}
+
+export async function getLatestBlockNumber(api: ApiPromise): Promise<number> {
+  let highHash = await api.rpc.chain.getFinalizedHead();
+  let high = (await api.rpc.chain.getHeader(highHash)).number.toNumber();
+  return high;
+}
+
+export async function getTimestampForBlockAt(api: ApiPromise, mid: number): Promise<number> {
+  const hash = await api.rpc.chain.getBlockHash(mid);
+  // FIXME: why is the conversion needed?
```

Comment:
TODO?

Reply:
Yeah, but I'm not really sure how to fix this. I get this error otherwise:

>  src/funnels/avail/utils.ts(37,34): error TS2345: Argument of type 'import("/home/enzo/Work/paima-engine/node_modules/avail-js-sdk/node_modules/@polkadot/types/interfaces/runtime/types").Header' is not assignable to parameter of type 'import("/home/enzo/Work/paima-engine/node_modules/@polkadot/types/interfaces/runtime/types").Header'.
>  Property '#private' is missing in type 'import("/home/enzo/Work/paima-engine/node_modules/avail-js-sdk/node_modules/@polkadot/types/interfaces/runtime/types").Header' but required in type 'import("/home/enzo/Work/paima-engine/node_modules/@polkadot/types/interfaces/runtime/types").Header'.

This probably happens because the types don't come from the same package (even if I think these are the same, since `avail-js-sdk` uses the polkadot package). But there is only one version of `@polkadot/types` in this package, so I'm not sure what's the conflict.